Fix: Calculation for occurrences include DST transitions correctly

axunonb opened the pull request:
Updated `CalendarEvent` and `EventEvaluator` to enhance event handling, particularly around time zones and daylight saving time transitions.

Key changes include:

- Enabled nullable reference types in `CalendarEvent` and `EventEvaluator.`
- Updated `CalendarEvent` to throw `InvalidOperationException` when trying to set both `DtEnd` and `Duration`.
- Event end time and duration calculations in `EventEvaluator` reflect any DST transitions.
- Added new test class `RecurrenceTests_From_Issues` to validate changes with various edge cases and scenarios.

Resolves #660
Resolves #623
Resolves #671
Resolves #567
Resolves #342
Resolves #298

axunonb committed:
Fix: Calculation for occurrences include DST transitions correctly

Updated `CalendarEvent` and `EventEvaluator` to enhance event handling,
particularly around time zones and daylight saving time transitions.

Key changes include:

- Enabled nullable reference types in `CalendarEvent` and `EventEvaluator.`
- Updated `CalendarEvent` to throw `InvalidOperationException` when trying to set both `DtEnd` and `Duration`.
- Event end time and duration calculations in `EventEvaluator` reflect any DST transitions.
- Added new test class `RecurrenceTests_From_Issues` to validate changes with various edge cases and scenarios.

Resolves #660
Resolves #623
Resolves #671
Resolves #567
Resolves #342
Resolves #298

minichma commented on Ical.Net/CalendarComponents/CalendarEvent.cs:
```diff
@@ -42,132 +54,136 @@ public class CalendarEvent : RecurringComponent, IAlarmContainer, IComparable<Ca
-    /// <returns>The effective duration of this event.</returns>
-    public virtual TimeSpan GetFirstDuration()
+    /// <returns>The <b>nominal duration</b> of this event.</returns>
+    public virtual TimeSpan CalcFirstNominalDuration()
```
Regarding the naming I'd rather go for just 'GetFirst...'. That calculations are done internally doesn't seem to be relevant from an outside perspective.

minichma replied:
If we improve the return value as suggested below, i.e. that it can represent the duration in a way that allows keeping nominal/exact apart, then we even wouldn't be talking about the 'first' duration anymore, so then a proper name could be like `GetEffectiveDuration()`. Like

`Duration GetEffectiveDuration()`

In `CalDateTime` we could then offer something like `Add(Duration d)` taking care of the difference between both parts.

axunonb replied:
See the [comment](https://github.com/ical-org/ical.net/pull/673#issuecomment-2558053226)

minichma commented on Ical.Net/CalendarComponents/CalendarEvent.cs:
```diff
@@ -42,132 +54,136 @@ public class CalendarEvent : RecurringComponent, IAlarmContainer, IComparable<Ca
-    /// <returns>The effective duration of this event.</returns>
-    public virtual TimeSpan GetFirstDuration()
+    /// <returns>The <b>nominal duration</b> of this event.</returns>
+    public virtual TimeSpan CalcFirstNominalDuration()
```
As long as we need to decide for *either* nominal *or* exact durations, I suggest going for exact, because if DTSTART and DTEND have different tzids, then there is no reasonable way to calculate a nominal duration. Nominal only makes sense in case of floating time or if both are in the same tz. If we just ignore the tz then the result might be significantly wrong, certainly more than when using exact time in cases where nominal would be correct. This is probably why the RFC specifies that the duration shall be considered exact if specivied via DTSTART/DTEND.

A better solution in a next step would probably be to return the duration in a way that allows the caller understand whether it should be applied as nominal or exact duration.

minichma replied:
Following up on [my comment in EventEvaluator](https://github.com/ical-org/ical.net/pull/673/files#r1893119377) - Got your point now of how the 'nominal' duration should be used, i.e. by requiring the caller to preserve DTEND's tz. However, this is a very specific case and I think we shouldn't offer this calculation as a built in functionality of CalendarEvent. If we (or anyone) needs this kind of calculation on times with different TZ, it should be calculated in place, exactly as its needed there, but not in such a central place. I'm not sure that 'nominal' is the right terminology if the TZ is different. Even if the terminology could be considered correct, its certainly very confusing and could be a source of coding flaws very easily (always talking just about the case with different tz). I therefore suggest to offer calculating the nominal duration *only* if the TZ is the same (or if the duration is specified via the 'DURATION' property'). Actually the only correct way of calculating the effective duration would be to calculate it in a way that preserves the info of whether it is nominal or exact rather than letting the caller choose whether they want nominal or exact.

This could be achieved in several ways:
1. Return something like `(TimeSpan nominal, TimeSpan exact)` that allows keeping both apart
2. Return something like `(TimeSpan d, bool isNominal)` - less expressive and not able to express a combination of both as required for the `DURATION` property.
3. Introduce a new `Duration` type that stores Days/Weeks and time independently, i.e. closer to how durations are represented in the RFC (`P1W2DT3H`). The RFC requires the days to be considered nominal while the time shall be considered exact. As we are free to return a time part > 24h, we would be free to represent `DTSTART - DTEND` just in the time part, which would imply its exact. This would require the Duration property of `CalendarEvent` to be represented as `Duration` rather than as `TimeSpan` either, which might be a bigger change.

I feel no 3 is probably the cleanest one while 1 is the most straight forward one.

axunonb replied:
I'd prefer option 3, too, but without replacing one intermediate implementation with another one.
See the [comment](https://github.com/ical-org/ical.net/pull/673#issuecomment-2558053226).

minichma commented on Ical.Net/CalendarComponents/CalendarEvent.cs:
```diff
@@ -42,132 +54,136 @@ public class CalendarEvent : RecurringComponent, IAlarmContainer, IComparable<Ca
+                same timezone as the event end time, finally leading to an exact duration
+                calculation from the zoned start time to the zoned end time.
+             */
+            return DtEnd.Value - DtStart.Value;
```
As mentioned above I suggest to go for the exact duration for now (i.e. `DtEnd.AsUtc - DtStart.AsUtc`). *If* calculating the nominal duration, we should throw if `DtEnd.TzId != DtStart.TzId`.

minichma commented on Ical.Net/Evaluation/EventEvaluator.cs:
```diff
@@ -27,37 +31,69 @@ public EventEvaluator(CalendarEvent evt) : base(evt) { }
+            // and use the timezone from the event's definition.
+            endTime = new CalDateTime(
+                DateOnly.FromDateTime(endDt),
+                period.StartTime.HasTime ? TimeOnly.FromDateTime(endDt) : null,
```
Here we risk confusing the tzid. Simply use `var endDt = period.StartTime.Add(nominalDuration)`? We must make sure that Add() and GetFirst..() both operate on the same kind of duration, nominal vs exact.

minichma replied:
Oh, I guess I missed your point. Of course, you apply the same 'nominal' duration (not sure whether this is the right term if the tz is different) and preserve the tz. So in the end the exact duration is close to the original one (except for the DST changes). Yeah, I guess this is a reasonable approach, but I would still discourage implementing it like that. The RFC requires applying the nominal duration *only* if specified via the `DURATION` property, in which case there is only a single tz. I guess this is due to the ambiguity that comes from comparing times in different zones. Anyhow, to my understanding there's no hint in the RFC of how to apply 'nominal' durations in case of different tz.

I therefore keep my suggestion to calculate in exact durations for now until we precisely implement keeping both apart (in which case we also don't have to deal with this ambiguity). This way we are not fully precise if a duration is specified via the `DURATION` property, but at lease if specified via `DTEND`. In the way it's implemented in the PR we could be off in both cases.

axunonb replied:
See [comment](https://github.com/ical-org/ical.net/pull/673#issuecomment-2558053226)

minichma commented on Ical.Net/Evaluation/EventEvaluator.cs:
```diff
@@ -27,37 +31,69 @@ public EventEvaluator(CalendarEvent evt) : base(evt) { }
+        }
+
+        // Return the Period object with the calculated end time and duration.
+        period.Duration = endTime.Subtract(period.StartTime); // exact duration
```
This might be the right time to clean up `Period` too, i.e. remove the redundant data and the `Extrapolate...` as we did in `CalendarEvent`.

axunonb replied:
Added a commit with Extrapolate... replaced

minichma commented on Ical.Net.Tests/CalendarEventTest.cs:
```diff
@@ -472,78 +475,99 @@ public void HourMinuteSecondOffsetParsingTest()
-            Assert.That(evt.GetFirstDuration(), Is.EqualTo(TimeSpan.FromHours(1)));
+            Assert.That(evt.DtStart.Value, Is.EqualTo(dt));
+            Assert.That(evt.DtEnd.Value, Is.EqualTo(dt.AddHours(1)));
+            Assert.That(evt.CalcFirstNominalDuration(), Is.EqualTo(TimeSpan.FromHours(1)));
```
It should not be possible to calculate a nominal duration if the tzids don't match as the value usually doesn't make much sense.

axunonb committed:
Merge branch 'main' into pr/first-event-duration

axunonb committed:
Refactor and improve Period class and related usages

`CalendarEvent`
- Renamed `CalcFirstNominalDuration` to `GetTimeSpanToAddToPeriodStartTime`. The logic remains unchanged.

`Period`
- Removed private `ExtrapolateTimes()` method
- Enabled nullable reference types
- Refactored constructors and properties
- Added `GetOriginalValues` method
- Enhanced documentation and exception handling
- Fixed bug in `CollidesWith(Period)`
- Added PeriodTests.cs to cover unit tests for `Period` class.
- Shortcoming: `Period.Duration` may still implicitly return a nominal or an exact duration.

minichma commented on Ical.Net/CalendarComponents/CalendarEvent.cs:
```diff
@@ -42,132 +54,137 @@ public class CalendarEvent : RecurringComponent, IAlarmContainer, IComparable<Ca
-    /// <returns>The effective duration of this event.</returns>
-    public virtual TimeSpan GetFirstDuration()
+    /// <returns>The time span that gets added to the period start time to get the period end time.</returns>
+    public virtual TimeSpan GetTimeSpanToAddToPeriodStartTime()
```
Should we maybe make this internal for now?

axunonb replied:
oh, yes, as this is only temporary

axunonb committed:
Make `CalendarEvent.GetTimeSpanToAddToPeriodStartTime()` internal
